### ISSUE TITLE
Use with Karma

### ISSUE THREAD

**Issue by Shukurdin-Aydarov** [pana-cc/mocha-typescript#112]
Hello! Сan I use mocha-typescript with karma? Thank you :)

**Comment by Shukurdin-Aydarov** [comment]
Yes, I have a test that run using Karma. Also I tried to run the test using mocha-typescript, but I have the following error: 
{
    "message": "Uncaught TypeError: Cannot set property 'mocha-typescript' of undefined\nat test/test.ts:9219:38\n\nTypeError: Cannot set property 'mocha-typescript' of undefined\n    at Object.<anonymous> (test/test.ts:9219:38)\n    at Object.<anonymous> (test/test.ts:9221:30)\n    at __webpack_require__ (test/test.ts:20:30)\n    at Object.defineProperty.value (test/test.ts:8542:28)\n    at __webpack_require__ (test/test.ts:20:30)\n    at module.exports (test/test.ts:84:18)\n    at test/test.ts:87:10",
    "str": "Uncaught TypeError: Cannot set property 'mocha-typescript' of undefined\nat test/test.ts:9219:38\n\nTypeError: Cannot set property 'mocha-typescript' of undefined\n    at Object.<anonymous> (test/test.ts:9219:38)\n    at Object.<anonymous> (test/test.ts:9221:30)\n    at __webpack_require__ (test/test.ts:20:30)\n    at Object.defineProperty.value (test/test.ts:8542:28)\n    at __webpack_require__ (test/test.ts:20:30)\n    at module.exports (test/test.ts:84:18)\n    at test/test.ts:87:10"
  } 
My simple test:
import { suite, test, slow, timeout } from "mocha-typescript";
@suite class Two {
    @test method() {}
}

I didn't try karma-typescript. I'm going to try now. Thank you!

**Comment by silkentrance** [comment]
@Shukurdin-Aydarov let us hear how this worked out. Perhaps you want to write a how to that we then can include in the documentation or you even make a PR for this yourself. 😄

**Comment by silkentrance** [comment]
This is related to #114.

**Comment by silkentrance** [comment]
@Shukurdin-Aydarov ping did you try out karma-typescript, how did it work out with mocha-typescript? Any news on this?

**Comment by silkentrance** [comment]
@Shukurdin-Aydarov ping

**Comment by jkoepcke** [comment]
@silkentrance maybe I can step in here - I see the same error as above with mocha-typescript. Using "plain" mocha works fine. What can I do to help?

**Comment by silkentrance** [comment]
@jkoepcke thanks for the offer. the problem, as I just found out is not necessarily that one cannot force karma to also include mocha-typescript in the build, e.g.

```
        files: [
            { pattern: "node_modules/expect.js/index.js" },
            { pattern: "node_modules/mocha-typescript/index.js" },
            { pattern: "src/**/*.ts" }
        ],
```

adding mocha-typescript to the list of files will show up other errors that are related to the way that mocha-typescript is being packaged.

First, mocha-typescript assumes that there is always an `exports` object, similarly so the typescript transpiler. Which of course fails when karma tries to package everything up and send it to the browser.

Second, the mocha-typescript module makes use of require(), as generated by the typescript transpiler, however, this is also not available.

So basically we must change the target module type to something that can be supported in the browser using karma.

Changing the target module to for example UMD or AMD will work and the so defined tests will not fail unless you force them to:

```
HeadlessChrome 71.0.3578 (Mac OS X 10.13.6) Suite notatest FAILED
	AssertionError: it works
	    at Suite.notatest (src/mts.spec.ts:10:16 <- src/mts.spec.js:18:21)
	    at Context.<anonymous> (/var/folders/g7/cl4q1rg90g74r5dcp6bbmghr0000gp/T/karma-typescript-bundle-6846ZjsTe86Yy6hQ.js:223:39)
HeadlessChrome 71.0.3578 (Mac OS X 10.13.6): Executed 2 of 2 (1 FAILED) (0.018 secs / 0.003 secs)
```

**Comment by silkentrance** [comment]
So what we would basically need are two different packages, one for use with karma or in the browser and one for use with node.

The main problem is that we use node specific stuff in our tests, so these would need different compiler options, target module commonjs instead of AMD/UMD.

**Comment by silkentrance** [comment]
The other problem is di/typedi.ts. When this gets compiled, the require for typedi will resolve to the same module, causing the error that Container is not defined in module typedi.

**Comment by silkentrance** [comment]
@pana-cc can we refactor out the di/typedi.ts module into a separate mocha-typescript-di-typedi  package? That way, users can add this as a test dependency and we would then also depend on typedi directly. as of now, typedi is just a dev dependency.

Of course, the registerDI function will still be included in mocha-typescript.

**Comment by silkentrance** [comment]
All work is done in #157.

**Comment by silkentrance** [comment]
@jkoepcke @Shukurdin-Aydarov  If you like, you can take a look at branch https://github.com/pana-cc/mocha-typescript/tree/gh-157. Here, we introduce the test/it/KarmaPackageIT integration test for the fixture located under test/it/fixtures/packages/karma-it, which is a simple project using karma, karma-typescript and so on. The tests in there will require mocha-typescript suite and test and work just fine.

**Comment by silkentrance** [comment]
ATM the build on travis will timeout as the package based tests take are very long time. I will have to refactor these out into separate integration test packages so that we can make use of lerna and it managing the dependencies.

**Comment by silkentrance** [comment]
Done on branch gh-157.
